File: util/net/http_transport_test_server.go
package main

import "bytes"
import "compress/gzip"
import "encoding/binary"
import "fmt"
import "io"
import "net/http"
import "net/http/httptest"
import "os"
import "strconv"
import "strings"

// consumeBodyAndDumpRequest is mostly from from net/http/httputil DumpRequest
// and slightly tweaked because we want the already-de-chunked de-gzip'd version
// than a slightly-more-raw version (mostly for compatibility with the previous
// Python version of this script).

func consumeBodyAndDumpRequest(req *http.Request) ([]byte, error) {
	var b bytes.Buffer

	// By default, print out the unmodified req.RequestURI, which
	// is always set for incoming server requests. But because we
	// previously used req.URL.RequestURI and the docs weren't
	// always so clear about when to use DumpRequest vs
	// DumpRequestOut, fall back to the old way if the caller
	// provides a non-server Request.
	reqURI := req.RequestURI
	if reqURI == "" {
		reqURI = req.URL.RequestURI()
	}

	fmt.Fprintf(&b, "%s %s HTTP/%d.%d\r\n", req.Method, reqURI, req.ProtoMajor, req.ProtoMinor)

	absRequestURI := strings.HasPrefix(req.RequestURI, "http://") || strings.HasPrefix(req.RequestURI, "https://")
	if !absRequestURI {
		host := req.Host
		if host == "" && req.URL != nil {
			host = req.URL.Host
		}
		if host != "" {
			fmt.Fprintf(&b, "Host: %s\r\n", host)
		}
	}

	if len(req.TransferEncoding) > 0 {
		fmt.Fprintf(&b, "Transfer-Encoding: %s\r\n", strings.Join(req.TransferEncoding, ","))
	}
	if req.Close {
		fmt.Fprintf(&b, "Connection: close\r\n")
	}

	var reqWriteExcludeHeaderDump = map[string]bool{
		"Host":              true, // not in Header map anyway
		"Transfer-Encoding": true,
		"Trailer":           true,
	}

	err := req.Header.WriteSubset(&b, reqWriteExcludeHeaderDump)
	if err != nil {
		return nil, err
	}

	io.WriteString(&b, "\r\n")

	if req.Body != nil {
		if req.Header.Get("Content-Encoding") == "gzip" {
			req.Body, err = gzip.NewReader(req.Body)
		}
		var dest io.Writer = &b
		_, err = io.Copy(dest, req.Body)
	}

	if err != nil {
		return nil, err
	}
	return b.Bytes(), nil
}

// A one-shot testing webserver.
//
// When invoked, this server will write a short integer to stdout, indicating on
// which port the server is listening. It will then read one short integer from
// stdin, indiciating the response code to be sent in response to a request. It
// also reads 16 characters from stdin, which, after having "\r\n" appended,
// will form the response body in a successful response (one with code 200). The
// server will process one HTTP request, deliver the prearranged response to the
// client, and write the entire request to stdout. It will then terminate.

func main() {
	var responseCode uint16
	var responseBody [16]byte
	var wholeRequest string
	done := make(chan int)
	ts := httptest.NewServer(
		http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			dump, err := consumeBodyAndDumpRequest(r)
			if err != nil {
				panic(err)
			}
			wholeRequest = string(dump[:len(dump)])

			if r.Method == "POST" && r.URL.Path == "/upload" {
				w.WriteHeader(int(responseCode))
				fmt.Fprint(w, string(responseBody[:len(responseBody)]), "\r\n")
				done <- 0
			} else {
				w.WriteHeader(500)
				done <- 1
			}
		}))

	parts := strings.Split(ts.URL, ":")
	port, err := strconv.Atoi(parts[len(parts)-1])
	if err != nil {
		panic(err)
	}
	binary.Write(os.Stdout, binary.LittleEndian, uint16(port))
	err = binary.Read(os.Stdin, binary.LittleEndian, &responseCode)
	if err != nil {
		panic(err)
	}
	binary.Read(os.Stdin, binary.LittleEndian, &responseBody)

	rc := <-done

	fmt.Fprint(os.Stdout, wholeRequest)

	ts.Close()
	os.Exit(rc)
}

File: util/net/http_transport_test_server.cc
#if 1
// Copyright 2018 The Crashpad Authors. All rights reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// A one-shot testing webserver.
//
// When invoked, this server will write a short integer to stdout, indicating on
// which port the server is listening. It will then read one integer from stdin,
// indicating the response code to be sent in response to a request. It also
// reads 16 characters from stdin, which, after having "\r\n" appended, will
// form the response body in a successful response (one with code 200). The
// server will process one HTTP request, deliver the prearranged response to the
// client, and write the entire request to stdout. It will then terminate.

#include "base/logging.h"
#include "base/strings/stringprintf.h"
#include "build/build_config.h"
#include "tools/tool_support.h"
#include "util/file/file_io.h"

#if COMPILER_MSVC
#pragma warning(push)
#pragma warning(disable: 4244 4245 4267 4702)
#endif

#define CPPHTTPLIB_ZLIB_SUPPORT
#include "third_party/cpp-httplib/cpp-httplib/httplib.h"

#if COMPILER_MSVC
#pragma warning(pop)
#endif

namespace crashpad {
namespace {

int HttpTransportTestServerMain(int argc, char* argv[]) {
  httplib::Server svr(httplib::HttpVersion::v1_0);

  if (!svr.is_valid()) {
    LOG(ERROR) << "server creation failed";
    return 1;
  }

  uint16_t response_code;
  char response[16];

  std::string to_stdout;

  svr.post("/upload",
           [&response, &response_code, &svr, &to_stdout](
               const httplib::Request& req, httplib::Response& res) {
             res.status = response_code;
             if (response_code == 200) {
               res.set_content(std::string(response, 16) + "\r\n",
                               "text/plain");
             } else {
               res.set_content("error", "text/plain");
             }

             for (const auto& h : req.headers) {
               fprintf(stderr,
                       "HEADER IN SERVER DUMP: '%s' => '%s'\n",
                       h.first.c_str(),
                       h.second.c_str());
               to_stdout += base::StringPrintf(
                   "%s: %s\r\n", h.first.c_str(), h.second.c_str());
             }
             to_stdout += "\r\n";
             to_stdout += req.body;

             svr.stop();
           });

  int port = svr.bind_to_any_port("127.0.0.1");

  CheckedWriteFile(
      StdioFileHandle(StdioStream::kStandardOutput), &port, sizeof(port));

  CheckedReadFileExactly(StdioFileHandle(StdioStream::kStandardInput),
                         &response_code,
                         sizeof(response_code));

  CheckedReadFileExactly(StdioFileHandle(StdioStream::kStandardInput),
                         &response,
                         sizeof(response));

  svr.listen_after_bind();

  fprintf(stderr, "WHOLE REQUEST DUMP FROM SERVER = '%s'\n", to_stdout.c_str());
  usleep(10000000);
  LoggingWriteFile(StdioFileHandle(StdioStream::kStandardOutput),
                   to_stdout.data(),
                   to_stdout.size());

  return 0;
}

} // namespace
} // namespace crashpad

#if defined(OS_POSIX) || defined(OS_FUCHSIA)
int main(int argc, char* argv[]) {
  return crashpad::HttpTransportTestServerMain(argc, argv);
}
#elif defined(OS_WIN)
int wmain(int argc, wchar_t* argv[]) {
  return crashpad::ToolSupport::Wmain(
      argc, argv, crashpad::HttpTransportTestServerMain);
}
#endif  // OS_POSIX

#else

//
//  sample.cc
//
//  Copyright (c) 2012 Yuji Hirose. All rights reserved.
//  The Boost Software License 1.0
//

#define CPPHTTPLIB_ZLIB_SUPPORT
#include "third_party/cpp-httplib/cpp-httplib/httplib.h"
#include <cstdio>
#include <chrono>

#define SERVER_CERT_FILE "./cert.pem"
#define SERVER_PRIVATE_KEY_FILE "./key.pem"

using namespace httplib;

std::string dump_headers(const Headers& headers)
{
    std::string s;
    char buf[BUFSIZ];

    for (auto it = headers.begin(); it != headers.end(); ++it) {
       const auto& x = *it;
       snprintf(buf, sizeof(buf), "%s: %s\n", x.first.c_str(), x.second.c_str());
       s += buf;
    }

    return s;
}

std::string log(const Request& req, const Response& res)
{
    std::string s;
    char buf[BUFSIZ];

    s += "================================\n";

    snprintf(buf, sizeof(buf), "%s %s %s", req.method.c_str(), req.path.c_str(), req.version.c_str());
    s += buf;

    std::string query;
    for (auto it = req.params.begin(); it != req.params.end(); ++it) {
       const auto& x = *it;
       snprintf(buf, sizeof(buf), "%c%s=%s",
           (it == req.params.begin()) ? '?' : '&', x.first.c_str(), x.second.c_str());
       query += buf;
    }
    snprintf(buf, sizeof(buf), "%s\n", query.c_str());
    s += buf;

    s += dump_headers(req.headers);

    s += "--------------------------------\n";

    snprintf(buf, sizeof(buf), "%d %s\n", res.status, res.version.c_str());
    s += buf;
    s += dump_headers(res.headers);
    s += "\n";

    if (!res.body.empty()) {
        s += res.body;
    }

    s += "\n";

    return s;
}

int main(void)
{
    auto version = httplib::HttpVersion::v1_0;

#ifdef CPPHTTPLIB_OPENSSL_SUPPORT
    SSLServer svr(SERVER_CERT_FILE, SERVER_PRIVATE_KEY_FILE, version);
#else
    Server svr(version);
#endif

    if (!svr.is_valid()) {
        printf("server has an error...\n");
        return -1;
    }

    svr.get("/", [=](const auto& /*req*/, auto& res) {
        res.set_redirect("/hi");
    });

    svr.get("/hi", [](const auto& /*req*/, auto& res) {
        res.set_content("Hello World!\n", "text/plain");
    });

    svr.get("/slow", [](const auto& /*req*/, auto& res) {
        using namespace std::chrono_literals;
        std::this_thread::sleep_for(2s);
        res.set_content("Slow...\n", "text/plain");
    });

    svr.get("/dump", [](const auto& req, auto& res) {
        res.set_content(dump_headers(req.headers), "text/plain");
    });

    svr.get("/stop", [&](const auto& /*req*/, auto& /*res*/) {
        svr.stop();
    });

    svr.set_error_handler([](const auto& /*req*/, auto& res) {
        const char* fmt = "<p>Error Status: <span style='color:red;'>%d</span></p>";
        char buf[BUFSIZ];
        snprintf(buf, sizeof(buf), fmt, res.status);
        res.set_content(buf, "text/html");
    });

    svr.set_logger([](const auto& req, const auto& res) {
        printf("%s", log(req, res).c_str());
    });

    int port = svr.bind_to_any_port("127.0.0.1");
    printf("on port: %d\n", port);
    svr.listen_after_bind();

    return 0;
}
// vim: et ts=4 sw=4 cin cino={1s ff=unix

#endif
